distsqlrun: make windower respect the memory limits

yuzefovich opened the pull request:
**distsqlrun: make windower respect the memory limits**

Previously, the windower didn't respect the memory limit testing knob and
the setting. Now this is fixed. There is one caveat though: windower
requires some amount of RAM to store its intermediate results, so if the
testing knob is lower than that, the limit is overwritten, but if the
cluster setting is insufficient, an error is returned.

Fixes: #40294.

**distsqlrun: make router output respect memory limit setting**

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/40340)
<!-- Reviewable:end -->

yuzefovich committed:
distsqlrun: make router output respect memory limit setting

Release note: None

jordanlewis reviewed (commented):
Generally <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but would like to get alfonso's review upon his return.

> Reviewed 10 of 10 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40340#-:-LnUpxwk0tmUx23UteDA:b-pgiwrk)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, @solongordon, and @yuzefovich)

___
*[pkg/sql/distsqlrun/server.go, line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40340#-LnUmxqbA1zkdzs1wxVj:-LnUmxqbA1zkdzs1wxVk:b-my36qi) ([raw file](https://github.com/cockroachdb/cockroach/blob/73df60ad4809182215f55b555f235fb96aacd899/pkg/sql/distsqlrun/server.go#L102)):*
> ```Go
> 	"set to true to enable use of disk for distributed sql window functions",
> 	true,
> )
> ```

I think that adding this variable is unnecessary. I don't see when it would be useful to turn it off, unless we use it in tests. I know we have the other ones, but I think we should delete them soon.
___
*[pkg/sql/distsqlrun/windower.go, line 278 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40340#-LnUpc6r4v3hoX87-j7s:-LnUpc6r4v3hoX87-j7t:b-qw3cdv) ([raw file](https://github.com/cockroachdb/cockroach/blob/73df60ad4809182215f55b555f235fb96aacd899/pkg/sql/distsqlrun/windower.go#L278)):*
> ```Go
> 		allRowsPartitioned := rowcontainer.MakeHashMemRowContainer(rc)
> 		w.allRowsPartitioned = &allRowsPartitioned
> 		if err := w.allRowsPartitioned.Init(
> ```

This seems to be identical in both branches of the if - is there some DRY to do here?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Thanks for the review!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40340#-:-LnUrWLm39YakSMTsikT:b-9qdoz3)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @solongordon)

___
*[pkg/sql/distsqlrun/server.go, line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40340#-LnUmxqbA1zkdzs1wxVj:-LnUriuD2H1fHNVQR9tA:b-gpbih9) ([raw file](https://github.com/cockroachdb/cockroach/blob/73df60ad4809182215f55b555f235fb96aacd899/pkg/sql/distsqlrun/server.go#L102)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I think that adding this variable is unnecessary. I don't see when it would be useful to turn it off, unless we use it in tests. I know we have the other ones, but I think we should delete them soon.
</blockquote></details>

Yeah, I wasn't sure whether the setting was useful or not and decided just to follow the pattern of joins and sorts.
___
*[pkg/sql/distsqlrun/windower.go, line 278 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40340#-LnUpc6r4v3hoX87-j7s:-LnUrKcmEzR8uhTugQ7P:bfrvxzh) ([raw file](https://github.com/cockroachdb/cockroach/blob/73df60ad4809182215f55b555f235fb96aacd899/pkg/sql/distsqlrun/windower.go#L278)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This seems to be identical in both branches of the if - is there some DRY to do here?
</blockquote></details>

There is a subtle difference - in the upper we create a disk backed container, and in the bottom one only an in-memory container.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40340)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, @solongordon, and @yuzefovich)

___
*[pkg/sql/distsqlrun/server.go, line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40340#-LnUmxqbA1zkdzs1wxVj:-Lnvyzlu1hihxcsiUNeE:bgtz628) ([raw file](https://github.com/cockroachdb/cockroach/blob/73df60ad4809182215f55b555f235fb96aacd899/pkg/sql/distsqlrun/server.go#L102)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Yeah, I wasn't sure whether the setting was useful or not and decided just to follow the pattern of joins and sorts.
</blockquote></details>

+1 I think we should remove the join and sort knobs too. I think wanting to exit rather than spilling to disk is a legit use case (maybe?) but that can be achieved by reducing the disk monitor limit (i.e. "oops I ran out of disk space"). This would also simplify the code.
___
*[pkg/sql/distsqlrun/windower.go, line 108 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40340#-LnvzMXTA67353p7Fgbc:-LnvzMXTA67353p7Fgbd:b-x0o677) ([raw file](https://github.com/cockroachdb/cockroach/blob/73df60ad4809182215f55b555f235fb96aacd899/pkg/sql/distsqlrun/windower.go#L108)):*
> ```Go
> // memRequiredByWindower indicates the minimum amount of RAM (in bytes) that
> // the windower needs.
> const memRequiredByWindower = 100000
> ```

Maybe easier to read with a multiplication (e.g. `100 * 1024`)
___
*[pkg/sql/distsqlrun/windower.go, line 278 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40340#-LnUpc6r4v3hoX87-j7s:-LnvzlwQ04unvw3NuTbX:b-z0pebu) ([raw file](https://github.com/cockroachdb/cockroach/blob/73df60ad4809182215f55b555f235fb96aacd899/pkg/sql/distsqlrun/windower.go#L278)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

There is a subtle difference - in the upper we create a disk backed container, and in the bottom one only an in-memory container.
</blockquote></details>

Can we remove this branch if we remove the cluster setting? The testing knob should enforce a hard limit so the only difference would be the monitor that the hash disk backed row container is created with.
___
*[pkg/sql/rowcontainer/hash_row_container.go, line 1011 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40340#-Lnw1VDi9ocviRIdNT6k:-Lnw1VDi9ocviRIdNT6l:b-9lp5op) ([raw file](https://github.com/cockroachdb/cockroach/blob/73df60ad4809182215f55b555f235fb96aacd899/pkg/sql/rowcontainer/hash_row_container.go#L1011)):*
> ```Go
> 	RowIterator
> 
> 	container HashRowContainer
> ```

Why is this changing to a `HashRowContainer`? I feel like intuitively we should only ever use spillable row containers. This seems like an artifact of the cluster setting's ability to turn off any disk spilling.


<!-- Sent from Reviewable.io -->

yuzefovich committed:
distsqlrun: make windower respect the memory limits

Previously, the windower didn't respect the memory limit testing knob and
the setting. Now this is fixed. There is one caveat though: windower
requires some amount of RAM to store its intermediate results, so if the
testing knob is lower than that, the limit is overwritten, but if the
cluster setting is insufficient, an error is returned.

Release note: None

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40340)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @jordanlewis, and @solongordon)

___
*[pkg/sql/distsqlrun/server.go, line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40340#-LnUmxqbA1zkdzs1wxVj:-LnxanHUCUez4yuLs0si:bku9sr3) ([raw file](https://github.com/cockroachdb/cockroach/blob/73df60ad4809182215f55b555f235fb96aacd899/pkg/sql/distsqlrun/server.go#L102)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

+1 I think we should remove the join and sort knobs too. I think wanting to exit rather than spilling to disk is a legit use case (maybe?) but that can be achieved by reducing the disk monitor limit (i.e. "oops I ran out of disk space"). This would also simplify the code.
</blockquote></details>

I agree. I removed the newly introduced setting and will remove others in the follow-up PR.
___
*[pkg/sql/distsqlrun/windower.go, line 108 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40340#-LnvzMXTA67353p7Fgbc:-Lnxax9O4yEGgAbwHRy0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/73df60ad4809182215f55b555f235fb96aacd899/pkg/sql/distsqlrun/windower.go#L108)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Maybe easier to read with a multiplication (e.g. `100 * 1024`)
</blockquote></details>

Done.
___
*[pkg/sql/distsqlrun/windower.go, line 278 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40340#-LnUpc6r4v3hoX87-j7s:-LnxazRi1Cby9xEjP6pH:bi7w4ia) ([raw file](https://github.com/cockroachdb/cockroach/blob/73df60ad4809182215f55b555f235fb96aacd899/pkg/sql/distsqlrun/windower.go#L278)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Can we remove this branch if we remove the cluster setting? The testing knob should enforce a hard limit so the only difference would be the monitor that the hash disk backed row container is created with.
</blockquote></details>

Refactored - removing the setting allowed for removing that branch.
___
*[pkg/sql/rowcontainer/hash_row_container.go, line 1011 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40340#-Lnw1VDi9ocviRIdNT6k:-Lnxb8Qc9a2k-v8MCTL0:b6ae7av) ([raw file](https://github.com/cockroachdb/cockroach/blob/73df60ad4809182215f55b555f235fb96aacd899/pkg/sql/rowcontainer/hash_row_container.go#L1011)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Why is this changing to a `HashRowContainer`? I feel like intuitively we should only ever use spillable row containers. This seems like an artifact of the cluster setting's ability to turn off any disk spilling.
</blockquote></details>

Exactly, that was only needed to support the setting. Removed.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 7 of 7 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40340#-:-LoK7019D_dYhe2Ms_8Y:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @jordanlewis, @solongordon, and @yuzefovich)

___
*[pkg/sql/distsqlrun/windower.go, line 209 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/40340#-LoK6URKEOeOxHkoAVjX:-LoK6URL876dwzYT2rJP:bfjcdc3) ([raw file](https://github.com/cockroachdb/cockroach/blob/13428eb52262265f9966b6db5e0b537dbcbdad60/pkg/sql/distsqlrun/windower.go#L209)):*
> ```Go
> 		processorID,
> 		output,
> 		nil, /* memMonitor */
> ```

Should this be nil rather than initialized after the monitor is determined? Might deserve a comment


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40340#-:-LoOCWHME1EItWVr9Uy1:b-tlx7ko)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @asubiotto, @jordanlewis, and @solongordon)

___
*[pkg/sql/distsqlrun/windower.go, line 209 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/40340#-LoK6URKEOeOxHkoAVjX:-LoOCFnF2oPtiKSSYIjt:b-sshl0z) ([raw file](https://github.com/cockroachdb/cockroach/blob/13428eb52262265f9966b6db5e0b537dbcbdad60/pkg/sql/distsqlrun/windower.go#L209)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Should this be nil rather than initialized after the monitor is determined? Might deserve a comment
</blockquote></details>

Good eye - I think I left as nil as a TODO but forgot to leave the TODO comment and actually refactor the code. Done.


<!-- Sent from Reviewable.io -->
